File: backend/src/main/java/com/m4hub/backend/service/ChatService.java
package com.m4hub.backend.service;

import com.m4hub.backend.model.ChatMessage;
import com.m4hub.backend.model.FriendRequest;
import com.m4hub.backend.model.User;
import com.m4hub.backend.repository.ChatMessageRepository;
import com.m4hub.backend.repository.FriendRequestRepository;
import com.m4hub.backend.repository.UserRepository;
import com.m4hub.backend.repository.MessageReactionRepository;
import com.m4hub.backend.repository.GroupChatRepository;
import com.m4hub.backend.repository.GroupMessageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class ChatService {

    private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ChatService.class);

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private FriendRequestRepository friendRequestRepository;

    @Autowired
    private ChatMessageRepository chatMessageRepository;

    @Autowired
    private MessageReactionRepository messageReactionRepository;

    @Autowired
    private GroupChatRepository groupChatRepository;

    @Autowired
    private GroupMessageRepository groupMessageRepository;

    @Autowired
    private SimpMessagingTemplate messagingTemplate;

    // --- Friend Requests ---

    public FriendRequest sendFriendRequest(Long senderId, String receiverUsername) {
        User sender = userRepository.findById(senderId).orElseThrow(() -> new RuntimeException("Sender not found"));
        User receiver = userRepository.findByUsername(receiverUsername)
                .orElseThrow(() -> new RuntimeException("User not found: " + receiverUsername));

        return createFriendRequest(sender, receiver);
    }

    public FriendRequest sendFriendRequestById(Long senderId, Long receiverId) {
        User sender = userRepository.findById(senderId).orElseThrow(() -> new RuntimeException("Sender not found"));
        User receiver = userRepository.findById(receiverId)
                .orElseThrow(() -> new RuntimeException("Receiver not found"));
        return createFriendRequest(sender, receiver);
    }

    private FriendRequest createFriendRequest(User sender, User receiver) {
        if (sender.getId().equals(receiver.getId())) {
            throw new RuntimeException("Cannot send request to yourself");
        }

        if (friendRequestRepository.existsByUsers(sender, receiver)) {
            throw new RuntimeException("Connection already exists or is pending");
        }

        FriendRequest request = new FriendRequest(sender, receiver, FriendRequest.Status.PENDING);
        FriendRequest saved = friendRequestRepository.save(request);

        // Notify the receiver via WebSocket
        try {
            messagingTemplate.convertAndSend("/queue/requests/" + receiver.getId(),
                    Map.of("type", "FRIEND_REQUEST", "action", "RELOAD"));
            logger.debug("Sent real-time notification to user ID: {}", receiver.getId());
        } catch (Exception e) {
            logger.error("Failed to send WebSocket notification", e);
        }

        return saved;
    }

    public List<FriendRequest> getPendingRequests(Long userId) {
        logger.debug("Fetching pending requests for user ID: {}", userId);
        List<FriendRequest> requests = friendRequestRepository.findByReceiverIdAndStatus(userId,
                FriendRequest.Status.PENDING);
        logger.debug("Found {} pending requests", requests.size());
        return requests;
    }

    public List<FriendRequest> getSentRequests(Long userId) {
        logger.debug("Fetching sent requests for user ID: {}", userId);
        List<FriendRequest> requests = friendRequestRepository.findBySenderIdAndStatus(userId,
                FriendRequest.Status.PENDING);
        logger.debug("Found {} sent requests", requests.size());
        return requests;
    }

    public void acceptRequest(Long requestId, Long userId) {
        FriendRequest request = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        if (!request.getReceiver().getId().equals(userId)) {
            throw new RuntimeException("Unauthorized");
        }

        request.setStatus(FriendRequest.Status.ACCEPTED);
        friendRequestRepository.save(request);
    }

    public void rejectRequest(Long requestId, Long userId) {
        FriendRequest request = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));

        if (!request.getReceiver().getId().equals(userId)) {
            throw new RuntimeException("Unauthorized");
        }

        request.setStatus(FriendRequest.Status.REJECTED);
        friendRequestRepository.save(request);
    }

    public List<User> getFriends(Long userId) {
        User user = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User not found"));
        List<FriendRequest> connections = friendRequestRepository.findAcceptedRequests(user);

        List<User> friends = new ArrayList<>();
        for (FriendRequest fr : connections) {
            if (fr.getSender().getId().equals(userId)) {
                friends.add(fr.getReceiver());
            } else {
                friends.add(fr.getSender());
            }
        }
        return friends;
    }

    // --- Chat Messages ---

    public ChatMessage sendMessage(Long senderId, Long receiverId, String content, String messageType,
            String mediaUrl) {
        User sender = userRepository.findById(senderId).orElseThrow(() -> new RuntimeException("Sender not found"));
        User receiver = userRepository.findById(receiverId)
                .orElseThrow(() -> new RuntimeException("Receiver not found"));

        ChatMessage message = new ChatMessage(sender, receiver, content);
        message.setMessageType(messageType);
        message.setMediaUrl(mediaUrl);
        ChatMessage saved = chatMessageRepository.save(message);

        // Broadcast real-time notifications
        try {
            Map<String, Object> payload = Map.of(
                    "id", saved.getId(),
                    "senderId", senderId,
                    "receiverId", receiverId,
                    "content", content,
                    "messageType", messageType,
                    "mediaUrl", mediaUrl != null ? mediaUrl : "",
                    "createdAt", saved.getCreatedAt().toString(),
                    "isRead", false,
                    "isDelivered", false);

            // Send to receiver
            messagingTemplate.convertAndSend("/queue/messages/" + receiverId, payload);
            // Send feedback to sender
            messagingTemplate.convertAndSend("/queue/messages/" + senderId, payload);

            logger.info("Broadcasted message from {} to {}", senderId, receiverId);
        } catch (Exception e) {
            logger.error("Failed to broadcast message via WebSocket", e);
        }

        return saved;
    }

    public List<ChatMessage> getConversation(Long userId, Long otherUserId) {
        User user1 = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User not found"));
        User user2 = userRepository.findById(otherUserId)
                .orElseThrow(() -> new RuntimeException("Other user not found"));

        List<ChatMessage> messages = chatMessageRepository.findConversation(user1, user2);

        // Mark as read if user is receiver
        /*
         * for (ChatMessage msg : messages) {
         * if (msg.getReceiver().getId().equals(userId) && !msg.isRead()) {
         * msg.setRead(true);
         * chatMessageRepository.save(msg);
         * }
         * }
         */
        return messages;
    }

    @Transactional
    public void markAsRead(Long userId, Long otherUserId) {
        User user1 = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User 1 not found"));
        User user2 = userRepository.findById(otherUserId).orElseThrow(() -> new RuntimeException("User 2 not found"));
        List<ChatMessage> messages = chatMessageRepository.findConversation(user1, user2);
        for (ChatMessage msg : messages) {
            if (msg.getReceiver().getId().equals(userId) && !msg.isRead()) {
                msg.setRead(true);
                chatMessageRepository.save(msg);
            }
        }
    }

    // Search users by username or name
    public List<User> searchUsers(String query) {
        return userRepository.searchUsers(query);
    }

    // Mark message as delivered
    @Transactional
    public void markAsDelivered(Long messageId) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message not found"));
        message.setDeliveredAt(java.time.Instant.now());
        chatMessageRepository.save(message);
    }

    // Mark message as read
    @Transactional
    public void markMessageAsRead(Long messageId) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message not found"));
        message.setReadAt(java.time.Instant.now());
        chatMessageRepository.save(message);
    }

    // Message Reactions
    @Transactional
    public void addReaction(Long messageId, Long userId, String emoji) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message not found"));
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Check if user already reacted
        Optional<com.m4hub.backend.model.MessageReaction> existing = messageReactionRepository
                .findByMessageAndUser(message, user);

        if (existing.isPresent()) {
            // Update existing reaction
            com.m4hub.backend.model.MessageReaction reaction = existing.get();
            reaction.setEmoji(emoji);
            messageReactionRepository.save(reaction);
        } else {
            // Create new reaction
            com.m4hub.backend.model.MessageReaction reaction = new com.m4hub.backend.model.MessageReaction(message,
                    user, emoji);
            messageReactionRepository.save(reaction);
        }
    }

    @Transactional
    public void removeReaction(Long messageId, Long userId) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message not found"));
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        messageReactionRepository.deleteByMessageAndUser(message, user);
    }

    public List<com.m4hub.backend.model.MessageReaction> getReactions(Long messageId) {
        ChatMessage message = chatMessageRepository.findById(messageId)
                .orElseThrow(() -> new RuntimeException("Message not found"));
        return messageReactionRepository.findByMessage(message);
    }

    // Group Chat
    @Transactional
    public com.m4hub.backend.model.GroupChat createGroup(Long creatorId, String name, String description) {
        User creator = userRepository.findById(creatorId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        com.m4hub.backend.model.GroupChat group = new com.m4hub.backend.model.GroupChat(name, creator);
        group.setDescription(description);
        return groupChatRepository.save(group);
    }

    @Transactional
    public void addGroupMember(Long groupId, Long userId, Long requesterId) {
        com.m4hub.backend.model.GroupChat group = groupChatRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Verify requester is a member
        User requester = userRepository.findById(requesterId)
                .orElseThrow(() -> new RuntimeException("Requester not found"));
        if (!group.getMembers().contains(requester)) {
            throw new RuntimeException("Unauthorized");
        }

        group.getMembers().add(user);
        groupChatRepository.save(group);
    }

    @Transactional
    public List<com.m4hub.backend.model.GroupChat> getUserGroups(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return groupChatRepository.findByMember(user);
    }

    @Transactional
    public void sendGroupMessage(Long groupId, Long senderId, String content, String messageType) {
        com.m4hub.backend.model.GroupChat group = groupChatRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));
        User sender = userRepository.findById(senderId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Verify sender is a member
        if (!group.getMembers().contains(sender)) {
            throw new RuntimeException("Unauthorized");
        }

        com.m4hub.backend.model.GroupMessage message = new com.m4hub.backend.model.GroupMessage(group, sender, content);
        message.setMessageType(messageType);
        com.m4hub.backend.model.GroupMessage saved = groupMessageRepository.save(message);

        group.setLastMessageAt(java.time.Instant.now());
        groupChatRepository.save(group);

        // Broadcast to each member of the group
        try {
            Map<String, Object> payload = Map.of(
                    "id", saved.getId(),
                    "groupId", groupId, // Still include groupId for clarity
                    "senderId", senderId,
                    "receiverId", groupId, // IMPORTANT: Frontend uses receiverId to match active group
                    "senderName", sender.getName() != null ? sender.getName() : sender.getUsername(),
                    "content", content,
                    "messageType", messageType,
                    "createdAt", saved.getCreatedAt().toString(),
                    "isGroup", true);

            // Send to ALL members' private queues
            for (User member : group.getMembers()) {
                messagingTemplate.convertAndSend("/queue/messages/" + member.getId(), payload);
            }
            logger.info("Broadcasted group message to {} members of group {}", group.getMembers().size(), groupId);
        } catch (Exception e) {
            logger.error("Failed to broadcast group message", e);
        }
    }

    public List<com.m4hub.backend.model.GroupMessage> getGroupMessages(Long groupId, Long userId) {
        com.m4hub.backend.model.GroupChat group = groupChatRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Verify user is a member
        if (!group.getMembers().contains(user)) {
            throw new RuntimeException("Unauthorized");
        }

        return groupMessageRepository.findByGroupOrderByCreatedAtAsc(group);
    }
}

File: mobile/services/chat.service.ts
import { Client, IMessage } from '@stomp/stompjs';
import SockJS from 'sockjs-client';
import axios from 'axios';

import 'text-encoding';
import { APP_CONFIG } from '../constants';

const API_URL = APP_CONFIG.API_URL;

export interface ChatMessage {
    id: number;
    senderId: number;
    receiverId: number;
    content: string;
    messageType: 'TEXT' | 'IMAGE' | 'VIDEO' | 'VOICE' | 'FILE';
    mediaUrl?: string;
    createdAt: string;
    deliveredAt?: string;
    readAt?: string;
    isDelivered: boolean;
    isRead: boolean;
}

export interface FriendRequest {
    id: number;
    sender: {
        id: number;
        username: string;
        name: string;
        email: string;
    };
    receiver: {
        id: number;
        username: string;
        name: string;
        email: string;
    };
    status: string;
    createdAt: string;
}

export interface UserSearchResult {
    id: number;
    username: string;
    name: string;
    email: string;
    isActive?: boolean;
}

class ChatService {
    private stompClient: Client | null = null;
    private messageCallbacks: ((message: ChatMessage) => void)[] = [];
    private typingCallbacks: ((data: { userId: number; isTyping: boolean }) => void)[] = [];
    private readCallbacks: ((data: { messageId: number; readAt: string }) => void)[] = [];
    private deliveryCallbacks: ((messageId: number) => void)[] = [];
    private requestCallbacks: (() => void)[] = [];
    private presenceCallbacks: ((data: { userId: number; isActive: boolean }) => void)[] = [];

    connect(userId: number, onConnected?: () => void) {
        if (this.stompClient?.active) return;

        this.stompClient = new Client({
            webSocketFactory: () => new SockJS(`${API_URL}/ws`),
            debug: (str) => console.log(str),
            reconnectDelay: 5000,
            heartbeatIncoming: 4000,
            heartbeatOutgoing: 4000,
            connectHeaders: {
                userId: userId.toString()
            }
        });

        this.stompClient.onConnect = () => {
            console.log('Mobile WebSocket Connected');

            this.stompClient?.subscribe(`/queue/messages/${userId}`, (message: IMessage) => {
                const chatMessage = JSON.parse(message.body) as ChatMessage;
                this.messageCallbacks.forEach(callback => callback(chatMessage));
            });

            this.stompClient?.subscribe(`/queue/typing/${userId}`, (message: IMessage) => {
                const data = JSON.parse(message.body);
                this.typingCallbacks.forEach(callback => callback(data));
            });

            this.stompClient?.subscribe(`/queue/read/${userId}`, (message: IMessage) => {
                const data = JSON.parse(message.body);
                this.readCallbacks.forEach(callback => callback(data));
            });

            this.stompClient?.subscribe(`/queue/delivered/${userId}`, (message: IMessage) => {
                const data = JSON.parse(message.body);
                this.deliveryCallbacks.forEach(callback => callback(data.messageId));
            });

            this.stompClient?.subscribe(`/queue/requests/${userId}`, (message: IMessage) => {
                console.log('Friend request notification received');
                this.requestCallbacks.forEach(callback => callback());
            });

            this.stompClient?.subscribe('/topic/presence', (message: IMessage) => {
                const data = JSON.parse(message.body);
                this.presenceCallbacks.forEach(callback => callback(data));
            });

            if (onConnected) onConnected();
        };

        this.stompClient.onStompError = (frame) => {
            console.error('Broker reported error: ' + frame.headers['message']);
            console.error('Additional details: ' + frame.body);
        };

        this.stompClient.activate();
    }

    disconnect() {
        if (this.stompClient) {
            this.stompClient.deactivate();
            this.stompClient = null;
        }
    }

    onMessage(callback: (message: ChatMessage) => void) {
        this.messageCallbacks.push(callback);
        return () => {
            this.messageCallbacks = this.messageCallbacks.filter(cb => cb !== callback);
        };
    }

    onTyping(callback: (data: { userId: number; isTyping: boolean }) => void) {
        this.typingCallbacks.push(callback);
        return () => {
            this.typingCallbacks = this.typingCallbacks.filter(cb => cb !== callback);
        };
    }

    onRead(callback: (data: { messageId: number; readAt: string }) => void) {
        this.readCallbacks.push(callback);
        return () => {
            this.readCallbacks = this.readCallbacks.filter(cb => cb !== callback);
        };
    }

    onDelivery(callback: (messageId: number) => void) {
        this.deliveryCallbacks.push(callback);
        return () => {
            this.deliveryCallbacks = this.deliveryCallbacks.filter(cb => cb !== callback);
        };
    }

    onRequest(callback: () => void) {
        this.requestCallbacks.push(callback);
        return () => {
            this.requestCallbacks = this.requestCallbacks.filter(cb => cb !== callback);
        };
    }

    onPresence(callback: (data: { userId: number; isActive: boolean }) => void) {
        this.presenceCallbacks.push(callback);
        return () => {
            this.presenceCallbacks = this.presenceCallbacks.filter(cb => cb !== callback);
        };
    }

    sendMessage(senderId: number, receiverId: number, content: string, messageType: string = 'TEXT', mediaUrl?: string) {
        if (this.stompClient && this.stompClient.active) {
            this.stompClient.publish({
                destination: '/app/chat.send',
                body: JSON.stringify({
                    senderId,
                    receiverId,
                    content,
                    messageType,
                    mediaUrl
                })
            });
        }
    }

    sendTypingIndicator(userId: number, receiverId: number, isTyping: boolean) {
        if (this.stompClient && this.stompClient.active) {
            this.stompClient.publish({
                destination: '/app/chat.typing',
                body: JSON.stringify({
                    userId,
                    receiverId,
                    isTyping
                })
            });
        }
    }

    markAsRead(messageId: number, senderId: number) {
        if (this.stompClient && this.stompClient.active) {
            this.stompClient.publish({
                destination: '/app/chat.read',
                body: JSON.stringify({
                    messageId,
                    senderId
                })
            });
        }
    }

    markAsDelivered(messageId: number) {
        if (this.stompClient && this.stompClient.active) {
            this.stompClient.publish({
                destination: '/app/chat.delivered',
                body: JSON.stringify({
                    messageId
                })
            });
        }
    }

    sendGroupMessage(senderId: number, groupId: number, content: string, messageType: string = 'TEXT') {
        if (this.stompClient && this.stompClient.active) {
            this.stompClient.publish({
                destination: '/app/group.send',
                body: JSON.stringify({
                    senderId,
                    groupId,
                    content,
                    messageType
                })
            });
        }
    }

    // REST API Calls (Token passed as argument for mobile)
    async sendFriendRequest(token: string, username?: string, userId?: number): Promise<any> {
        const payload: any = {};
        if (userId) payload.userId = userId;
        else if (username) payload.username = username;

        const response = await axios.post(
            `${API_URL}/api/chat/request/send`,
            payload,
            { headers: { Authorization: `Bearer ${token}` } }
        );
        return response.data;
    }

    async getPendingRequests(token: string): Promise<FriendRequest[]> {
        const response = await axios.get(`${API_URL}/api/chat/requests/pending`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    async getSentRequests(token: string): Promise<FriendRequest[]> {
        const response = await axios.get(`${API_URL}/api/chat/requests/sent`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    async acceptRequest(requestId: number, token: string): Promise<any> {
        const response = await axios.post(
            `${API_URL}/api/chat/request/accept`,
            { requestId },
            { headers: { Authorization: `Bearer ${token}` } }
        );
        return response.data;
    }

    async rejectRequest(requestId: number, token: string): Promise<any> {
        const response = await axios.post(
            `${API_URL}/api/chat/request/reject`,
            { requestId },
            { headers: { Authorization: `Bearer ${token}` } }
        );
        return response.data;
    }

    async getFriends(token: string): Promise<UserSearchResult[]> {
        const response = await axios.get(`${API_URL}/api/chat/friends`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    async getConversation(otherUserId: number, token: string): Promise<ChatMessage[]> {
        const response = await axios.get(`${API_URL}/api/chat/conversation/${otherUserId}`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    async searchUsers(query: string, token: string): Promise<UserSearchResult[]> {
        const response = await axios.get(`${API_URL}/api/chat/search`, {
            params: { query },
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    async addReaction(messageId: number, emoji: string, token: string): Promise<any> {
        const response = await axios.post(
            `${API_URL}/api/chat/message/${messageId}/reaction`,
            { emoji },
            { headers: { Authorization: `Bearer ${token}` } }
        );
        return response.data;
    }

    async createGroup(name: string, description: string, token: string, memberIds: number[] = []): Promise<any> {
        const response = await axios.post(
            `${API_URL}/api/chat/group/create`,
            { name, description, memberIds },
            { headers: { Authorization: `Bearer ${token}` } }
        );
        return response.data;
    }

    async getGroups(token: string): Promise<any[]> {
        const response = await axios.get(`${API_URL}/api/chat/groups`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    async addGroupMember(groupId: number, userId: number, token: string): Promise<any> {
        const response = await axios.post(
            `${API_URL}/api/chat/group/${groupId}/members`,
            { userId },
            { headers: { Authorization: `Bearer ${token}` } }
        );
        return response.data;
    }

    async getGroupMessages(groupId: number, token: string): Promise<any[]> {
        const response = await axios.get(`${API_URL}/api/chat/group/${groupId}/messages`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }

    // Get relationship status with a user
    async getUserRelationshipStatus(userId: number, token: string): Promise<{
        isFriend: boolean;
        isPending: boolean; // User sent request to me
        isSent: boolean; // I sent request to user
    }> {
        try {
            const [friends, pending, sent] = await Promise.all([
                this.getFriends(token),
                this.getPendingRequests(token),
                this.getSentRequests(token)
            ]);

            return {
                isFriend: friends.some(f => f.id === userId),
                isPending: pending.some(r => r.sender.id === userId),
                isSent: sent.some(r => r.receiver.id === userId)
            };
        } catch (error) {
            console.error('Error getting relationship status:', error);
            return { isFriend: false, isPending: false, isSent: false };
        }
    }
}

export default new ChatService();
